internal/core/software_renderer.rs

```diff
@@ -1994,15 +1994,7 @@ impl<'a, T: ProcessScene> crate::item_rendering::ItemRenderer for SceneBuilder<'
         let color = self.alpha_color(text_input.color().color());
         let max_size = (geom.size.cast() * self.scale_factor).cast();
 
-        // Clip glyphs not only against the global clip but also against the Text's geometry to avoid drawing outside
-        // of its boundaries (that breaks partial rendering and the cast to usize for the item relative coordinate below).
```

Comment:
> I wonder if the proper fix is not to keep the cursor inside the bounds.

That seems sensible to me. I bet though that with other renderers the same issue is reproducible (if they'd use the partial renderer logic). The cursor width is configurable, so with a wide enough cursor it would be drawn out of (visual) bounds.

Perhaps we're "holding it wrong" and the cursor should be drawn separately from the text. After all, imagine a large paragraph of text with a blinking cursor: We don't want to re-render the entire text every time the cursor blinks. Ideally only the region with the cursor, as a compromise perhaps just the line of text. The rest should be clipped away.

Reply:
Yes exactly. Technically it’s only needed when the text input has focus, so it’s also not that bad.

Reply:
but the memory overhead will be paid always.
Unless we make it
```
TextInput {
   //...    

    // inserted by compiler
    if self.cursor-visible: Rectangle {
       background: parent.color;
       width: parent.text-cursor-width;
       height: parent.font-size;
       // cursor_position is a new property
       x: parent.cursor_position.x;
       y: parent.cursor_position.y;
    }
}
```

But there is still some overhead, and even more overhead when it has the focus.